docs/ccv1_image_security_design.md

```diff
@@ -0,0 +1,345 @@
+        "creator": "some software package v1.0.1-35",
+        "timestamp": 1483228800,
+    }
+}
+```
+
+The process of signing a container image is as follows:
+
+1. Create a signature JSON payload and write the manifest digest together with some otheridentity information of the container image.
+2. Use the private key to sign the JSON payload generated in the previous step, and package the cryptographic signature together with the JSON payload into a signature claim file.
+
+The signature claim file is what we usually call the container image signature file. Its directory structure and file name format are as follows:
+
+```
+.
+└── busybox@sha256=817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
+    └── signature-1
+```
+
+The meaning of directory name is `image-name@digest-algo=digest-value`, multiple signature files can exist in the directory, which means that the same container image can have multiple signatures and be numbered and stored.
+
+### Image distribution
+
+After the encryption and signing are completed, we obtain two contents: the container image itself and the signature claim file, which are distributed and stored separately.
+
+- container image: pushed to the specified remote registry.
+
+- signature claim file: upload to the specified sigstore server.
+
+## Deployment of protected container image
+
+In the confidential container V1 design, security functions will be performed when pulling the container image, including checking the registry whitelist, verifying the container image signature, and decrypting the container image layer. The high-level process is as follows:
+
+1. Check whether the registry where the image is located is in policy.json's whitelist.
+
+2. Download the signature claim file of the container image according to the configuration of the sigstore file.
+
+3. Verify the signature of the container image according to the configuration of the policy.json file.
+
+4. Pull the container image layer by layer. If an encrypted container image layer is encrypted, decrypt it.
+
+After the image is pulled, the subsequent deployment process is consistent with that of the normal container image.
+
+### policy.json
+
+policy.json is a necessary configuration file for deploying images in the confidential container V1 design. Its location is specified in the /etc/containers/ directory of boot image rootfs. Its main functions are as follows:
+
+- Declare a whitelist of container images with registry granularity
+
+- Configure the public key to use when verifying the image signature
+
+Thus, policy.json is confidential information that must be controlled by users and cannot be read or modified by untrusted entities.
+
+If image-rs finds that there is no policy.json in the /etc/containers/ directory or If the policy.json file is empty, call the GetResource grpc service provided by the attestation-agent. After remote attestation, obtain it from the user controlled relying KBS through the encrypted channel.
```

Comment:
In the mode described in the document, image-rs will preferentially read the configuration file under the local specified directory. If it is found that the config file does not exist, it will get it from AA. In the scenario where a trusted rootfs is provided, the configuration file in the local specified directory will be preset in advance, so the service interface of AA will not be called to obtain it. In this way, it can be compatible with the scenarios of trusted rootfs and untrusted rootfs

Do you mean that we need an explicit configuration option to configure whether image-rs is obtained directly from the local specified directory or from the service interface of AA?

Reply:
@jialez0 Do you see it as a strict necessity that `policy.json` cannot be _read_ by untrusted entities? I fully understand they should not be able to _write_ it, but do we consider a list of allowed container images plus a public key for verification to be confidential data? I wouldn't understand what the threat is there.

Reply:
> Do you mean that we need an explicit configuration option to configure whether image-rs is obtained directly from the local specified directory or from the service interface of AA?

No I think it's good as it is. If someone doesn't want to use signatures they can just include the permissive config that you describe below in their initrd. This adds additional overhead for the basic user who just wants to deploy containers, but I guess it's fine.

Reply:
@Jakob-Naucke At present, ccv1, there are two distribution methods for policy files. In the protected boot image scheme, the user directly presets them in the rootfs of the boot image. In the unprotected boot image scheme, they are dynamically obtained from kbs through AA. KBS will distribute the policy only after AA's attitude evidence is verified, Sending through the encrypted channel is not to prevent untrusted entities from reading him, but to prevent malicious tampering during transmission. Of course, only Mac can be used to protect the integrity of policy files, which depends on the security requirements of specific KBC and KBS.

Reply:
> Sending through the encrypted channel is not to prevent untrusted entities from reading him, but to prevent malicious tampering during transmission.

Shouldn't it be this then?
```
Thus, policy.json is confidential information that must be controlled by users and cannot be modified by untrusted entities.
```

Reply:
> 

My statement here may not be accurate and should be modified as follows: `The policy file must be generated and maintained by the user. Additional security means need to be used to ensure that the policy file will not be modified by malicious entities during storage and transmission. If users have high requirements for security level, they even need to use additional means to ensure the confidentiality of policy files.`

Reply:
The implication of the above is that a local rootfs `policy.json` takes precedence over a remote AA one. This seems wrong to me. Consider if some particular registry has been compromised or simply moved/renamed, but the rest of your encrypted images are otherwise good (possibly at a new location). Wouldn't it be preferable to add a `policy.json` to your KBS that would supersede the local one, rather than force all rootfs to be re-generated?